crates/txpool/src/lib.rs

```diff
@@ -1,10 +1,116 @@
+//!
+//! NOTE: the transactions from a single sender are not ordered by priority,
+//! but still when constructing pending set we always need to maintain the ordering
+//! (i.e. `txs[1]` always needs to be included after `txs[0]` even if it has higher priority)
+//!
+//! ### Design Details
+//!
+//! Performance assumptions:
+//! - Possibility to handle tens of thousands of transactions
+//! - Fast insertions and replacements `O(per-sender + log(senders))`
+//! - Reasonably fast removal of stalled transactions `O(per-sender)`
+//! - Reasonably fast construction of pending set `O(txs * (log(senders) + log(per-sender))`
+//!
+//! The removal performance could be improved by trading some memory. Currently `SmallVec` is used
+//! to store senders transactions, instead we could use `VecDeque` and efficiently `pop_front`
+//! the best transactions.
+//!
+//! The pending set construction and insertion complexity could be reduced by introducing
+//! a notion of `nonce` - an absolute, numeric ordering of transactions.
+//! We don't do that because of possible implications of EIP208 where nonce might not be
+//! explicitly available.
+//!
+//! 1. The pool groups transactions from particular sender together
+//!    and stores them ordered by `Scoring` within that group
+//!    i.e. `HashMap<Sender, Vec<Transaction>>`.
+//! 2. Additionaly we maintain the best and the worst transaction from each sender
+//!    (by `Scoring` not `priority`) ordered by `priority`.
+//!    It means that we can easily identify the best transaction inside the entire pool
+//!    and the worst transaction.
+//! 3. Whenever new transaction is inserted to the queue:
+//!    - first check all the limits (overall, memory, per-sender)
+//!    - retrieve all transactions from a sender
+//!    - binary search for position to insert the transaction
+//!    - decide if we are replacing existing transaction (3 outcomes: drop, replace, insert)
+//!    - update best and worst transaction from that sender if affected
+//! 4. Pending List construction:
+//!    - Take the best transaction (by priority) from all senders to the List
+//!    - Replace the transaction with next transaction (by ordering) from that sender (if any)
+//!    - Repeat
+
+#![warn(missing_docs)]
 
 #[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
-    }
+mod tests;
+
+mod error;
+mod listener;
+mod options;
+mod pool;
+mod ready;
+mod replace;
+mod status;
+mod transactions;
+mod verifier;
+
+pub mod scoring;
+
+pub use self::{
+    error::Error,
+    listener::{Listener, NoopListener},
+    options::Options,
+    pool::{PendingIterator, Pool, Transaction, UnorderedIterator},
+    ready::{Readiness, Ready},
+    replace::{ReplaceTransaction, ShouldReplace},
+    scoring::Scoring,
+    status::{LightStatus, Status},
+    verifier::Verifier,
+};
+
+use std::{fmt, hash::Hash};
+
+/// Already verified transaction that can be safely queued.
+pub trait VerifiedTransaction: fmt::Debug {
```

Comment:
Perhaps we can mark verified transactions on the container level rether than the type level? i.e. having a colleciton of `verified_transactions: Vec<Transaction>` and `unverified_transactions: Vec<Transaction>`?